From 9168e236c548d1d0e9938ee6dd4cdbd308fdfd72 Mon Sep 17 00:00:00 2001
From: Nikita Popov <nikita.ppv@gmail.com>
Date: Sun, 13 Feb 2022 14:13:16 +0100
Subject: [PATCH 1013/1013] [rust] Declare madvise() on illumos

Illumos seems to have made a real mess here. Adopt the same
workaround as https://reviews.llvm.org/D84046 and
https://reviews.llvm.org/D86166.
---
 llvm/lib/Support/Unix/Path.inc | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/llvm/lib/Support/Unix/Path.inc b/llvm/lib/Support/Unix/Path.inc
index 85352b8d68ef..c58432bbb030 100644
--- a/llvm/lib/Support/Unix/Path.inc
+++ b/llvm/lib/Support/Unix/Path.inc
@@ -118,6 +118,12 @@ typedef uint_t uint;
 #define STATVFS_F_FLAG(vfs) (vfs).f_flags
 #endif
 
+#if defined(__sun__)
+// The madvise() declaration on Illumos cannot be made visible if _XOPEN_SOURCE
+// is defined. This declaration is also compatible with Solaris 11.4.
+extern "C" int madvise(void *, size_t, int);
+#endif
+
 using namespace llvm;
 
 namespace llvm {
-- 
2.35.1

